docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,122 @@
+- limits
+    - a validator cannot rotate its consensus key more than `MaxConsPubKeyRotations` time for any unbonding period, to prevent spam.
+    - parameters can be decided by governance and stored in genesis file.
+- key rotation fee
+    - a validator should pay `KeyRotationFee` to rotate the consensus key which is calculated as below
+    - `KeyRotationFee` = `InitialKeyRotationFee` * 2^(number of rotations in `ConsPubKeyRotationHistory` in recent unbonding period)
+- evidence module
+    - evidence module can search corresponding consensus key for any height from slashing keeper so that it can decide which consensus key is supposed to be used for given height.
+- abci.ValidatorUpdate
+    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
+    - validator consensus key update can be done via creating new + delete old by change the power to zero.
+    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.
+- new genesis parameters in `staking` module
+    - `MaxConsPubKeyRotations` : maximum number of rotation can be executed by a validator in recent unbonding period. default value 10 is suggested(11th key rotation will be rejected)
+    - `InitialKeyRotationFee` : the initial key rotation fee when no key rotation has happened in recent unbonding period. default value 1atom is suggested(1atom fee for the first key rotation in recent unbonding period)
+
+
+### Workflow
+
+1. The validator generates a new consensus keypair.
+2. The validator generates and signs a `MsgRotateConsPubKey` tx with their operator key and new ConsPubKey
+
+    ```go
+    type MsgRotateConsPubKey struct {
+        ValidatorAddress  sdk.ValAddress
+        NewPubKey         crypto.PubKey
+    }
+    ```
+
+3. `handleMsgRotateConsPubKey` gets `MsgRotateConsPubKey`, calls `RotateConsPubKey` with emits event
+4. `RotateConsPubKey` 
+    - checks if `NewPubKey` is not duplicated on `ValidatorsByConsAddr`
+    - checks if the validator is does not exceed parameter `MaxConsPubKeyRotations` by iterating `ConsPubKeyRotationHistory`
+    - checks if the signing account has enough balance to pay `KeyRotationFee`
+    - pays `KeyRotationFee` to community fund
+    - overwrites `NewPubKey` in `validator.ConsPubKey`
+    - deletes old `ValidatorByConsAddr`
+    - `SetValidatorByConsAddr` for `NewPubKey`
+    - Add `ConsPubKeyRotationHistory` for tracking rotation
+
+    ```go
+    type ConsPubKeyRotationHistory struct {
+        OperatorAddress         sdk.ValAddress
+        OldConsPubKey           crypto.PubKey
+        NewConsPubKey           crypto.PubKey
+        RotatedHeight           int64
+    }
+    ```
+
+5. `ApplyAndReturnValidatorSetUpdates` checks if there is `ConsPubKeyRotationHistory` with `ConsPubKeyRotationHistory.RotatedHeight == ctx.BlockHeight()` and if so, generates 2 `ValidatorUpdate` , one for a remove validator and one for create new validator 
```

Comment:
Will this change the priority of the validator in the RoundRobin proposer selection in Tendermint?

Reply:
The new rotated key will have the lowest accum because it is seen as a new validator. So it is a slight disadvantage to the key rotated validator